qctrlopencontrols/pyquil/program.py

```diff
@@ -0,0 +1,209 @@
+    pulses are filled up with `identity` gates. Each identity gate introduces a delay of
+    `gate_time`. In this implementation, the number of identity gates is determined by
+    :math:`np.int(np.floor(offset_distance / gate_time))`. As a consequence, the duration of
+    the real-circuit is :math:`gate_time \\times number_of_identity_gates +
+    pulse_gate_time \\times number_of_pulses`.
+
+    Q-CTRL Open Controls support operation resulting in rotation around at most one axis at
+    any offset.
+    """
+
+    if dynamic_decoupling_sequence is None:
+        raise ArgumentsValueError('No dynamic decoupling sequence provided.',
+                                  {'dynamic_decoupling_sequence': dynamic_decoupling_sequence})
+
+    if not isinstance(dynamic_decoupling_sequence, DynamicDecouplingSequence):
+        raise ArgumentsValueError('Dynamical decoupling sequence is not recognized.'
+                                  'Expected DynamicDecouplingSequence instance',
+                                  {'type(dynamic_decoupling_sequence)':
+                                       type(dynamic_decoupling_sequence)})
+
+    if target_qubits is None:
+        target_qubits = [0]
+
+    if gate_time <= 0:
+        raise ArgumentsValueError(
+            'Time delay of identity gate must be greater than zero.',
+            {'gate_time': gate_time})
+
+    if np.any(target_qubits) < 0:
+        raise ArgumentsValueError(
+            'Every target qubits index must be positive.',
+            {'target_qubits': target_qubits})
+
+    if algorithm not in [FIX_DURATION_UNITARY, INSTANT_UNITARY]:
+        raise ArgumentsValueError('Algorithm must be one of {} or {}'.format(
+            INSTANT_UNITARY, FIX_DURATION_UNITARY), {'algorithm': algorithm})
+
+    unitary_time = 0.
+    if algorithm == FIX_DURATION_UNITARY:
+        unitary_time = gate_time
+
+    rabi_rotations = dynamic_decoupling_sequence.rabi_rotations
+    azimuthal_angles = dynamic_decoupling_sequence.azimuthal_angles
+    detuning_rotations = dynamic_decoupling_sequence.detuning_rotations
+
+    if len(rabi_rotations.shape) == 1:
+        rabi_rotations = rabi_rotations[np.newaxis, :]
+    if len(azimuthal_angles.shape) == 1:
+        azimuthal_angles = azimuthal_angles[np.newaxis, :]
+    if len(detuning_rotations.shape) == 1:
+        detuning_rotations = detuning_rotations[np.newaxis, :]
+
+    operations = np.vstack((rabi_rotations, azimuthal_angles, detuning_rotations))
+    offsets = dynamic_decoupling_sequence.offsets
+
+    time_covered = 0
+    program = Program()
+    program += Pragma('PRESERVE_BLOCK')
+
+    for operation_idx in range(operations.shape[1]):
+
+        offset_distance = offsets[operation_idx] - time_covered
+
+        if np.isclose(offset_distance, 0.0):
+            offset_distance = 0.0
+
+        if offset_distance < 0:
+            raise ArgumentsValueError("Offsets cannot be placed properly",
```

Comment:
That is valid point; However, Open-Controls for now only has `ArgumentsValueError`

Reply:
Ah, didn't realise that. Let's wait and see what Michael says, but I don't think we need to block this PR on that (since we'll need to make changes across the board).

Reply:
I think ArgumentsValueError is right, the DDS is the problem here. But:
1) The error message is not clear enough, you should articulate specifically what the problem with the DDS is, in a manner that can help figure out how to fix it. It appears this occurs when the spacing of the gates is smaller than the gate block time? Is that right?
2) The dictionary returned with the ArgumentsValueError should always be a list of strings with the exact name of the argument that is a problem. So in this case it should at least be the dynamical_decoupling_sequence. The extras keyword can be used to provide a dictionary with other additional useful information.
3) Why are you taking a str? You should be returning the object itself in the Arguments Value Error. If it's to avoid printing the repr rather than the str, I think the repr should be printed, it's typically more informative for diagnosing errors.

Reply:
You're right. Somehow I'd thought that we were generating the DDS internally in this method.